### PR TITLE
Fix vae

### PR DESCRIPTION
**Pull request by jheek**

### CODE REVIEW

File: examples/vae/main.py [hunk 1]
```diff
@@ -142,30 +139,37 @@ def eval(model, eval_ds, z):
...
 
   optimizer = optim.Adam(learning_rate=FLAGS.learning_rate).create(vae)
 
+  rng, z_key, eval_rng = random.split(rng, 3)
+  z = random.normal(z_key, (64, 20))
+
   for epoch in range(FLAGS.num_epochs):
     for batch in tfds.as_numpy(train_ds):
+      rng, key = random.split(rng)
       batch['image'] = batch['image'].reshape(-1, 784) / 255.0
```
Comment (AlexeyG) [inline examples/vae/main.py]: Why are images flattened and  then unflattened again in `eval`? If flattening serves a purpose, consider moving the first flattening op into the IO pipeline

File: examples/vae/main.py [hunk 2]
```diff
@@ -142,30 +139,37 @@ def eval(model, eval_ds, z):
...
   vae = nn.Model(VAE, params)
 
   optimizer = optim.Adam(learning_rate=FLAGS.learning_rate).create(vae)
 
+  rng, z_key, eval_rng = random.split(rng, 3)
+  z = random.normal(z_key, (64, 20))
+
   for epoch in range(FLAGS.num_epochs):
     for batch in tfds.as_numpy(train_ds):
+      rng, key = random.split(rng)
```
Comment (AlexeyG) [inline examples/vae/main.py]: What's the reason for manually handling key splitting instead of using the stochastic context? Using the latter would showcase more of Flax's features.

File: examples/vae/main.py (hunk 1 above)
Comment (AlexeyG) [inline examples/vae/main.py]: After dividing by 255. the data becomes "continuous". You need to dynamically binarise it- both for train and eval. Many VAEs report results on dynamically binarised MNIST and on statically binarised MNIST (some fixed dataset, which I believe isn't a part of TFDS).

File: examples/vae/main.py [hunk 3]
```diff
@@ -142,30 +139,37 @@ def eval(model, eval_ds, z):
...
+  rng = random.PRNGKey(0)
+  rng, key = random.split(rng)
+
   train_ds = tfds.load('mnist', split=tfds.Split.TRAIN)
   train_ds = train_ds.cache().shuffle(1000).batch(FLAGS.batch_size)
   test_ds = tfds.as_numpy(tfds.load('mnist', split=tfds.Split.TEST, batch_size=-1))
 
-  _, params = VAE.init_by_shape(key, [((1, 784), jnp.float32)])
+  _, params = VAE.init_by_shape(
+      key, [((1, 784), jnp.float32)], z_rng=random.PRNGKey(0))
```
Comment (AlexeyG) [inline examples/vae/main.py]: It doesn't make a difference for this specific model, but I think it'd be cleaner to specify the actual batch dimension during model init (instead of dim 1 as it is now).

File: examples/vae/README.md [hunk 4]
```diff
@@ -12,6 +12,10 @@ python main.py
```
Comment (AlexeyG) [inline examples/vae/README.md]: Git doesn't let me comment on parts that are not considered different. But could you please remove a semicolon after `## Examples` ?

File: examples/vae/main.py [hunk 5]
```diff
@@ -142,30 +139,37 @@ def eval(model, eval_ds, z):
...
   vae = nn.Model(VAE, params)
 
   optimizer = optim.Adam(learning_rate=FLAGS.learning_rate).create(vae)
 
+  rng, z_key, eval_rng = random.split(rng, 3)
+  z = random.normal(z_key, (64, 20))
```
Comment (AlexeyG) [inline examples/vae/main.py]: 64 -> FLAGS.batch_size ?

Comment (AlexeyG) [reply examples/vae/main.py]: Can we also make 20 (latent dimensionality) a parameter? I don't expect it being used a lot, but giving it a name would improve readability IMO.

File: examples/vae/main.py [hunk 6]
```diff
@@ -127,10 +124,10 @@ def loss_fn(model):
...
 
 @jax.jit
-def eval(model, eval_ds, z):
+def eval(model, eval_ds, z, z_rng):
   xs = eval_ds['image'] / 255.0
   xs = xs.reshape(-1, 784)
-  recon_xs, mean, logvar = model(xs)
+  recon_xs, mean, logvar = model(xs, z_rng)
 
   comparison = jnp.concatenate([xs[:8].reshape(-1, 28, 28, 1),
```
Comment (AlexeyG) [inline examples/vae/main.py]: Return all samples+reconstructions and subsample+concat outside of eval before saving the image?

File: examples/vae/main.py (hunk 1 above)
Comment (AlexeyG) [reply examples/vae/main.py]: Scrap half of the previous comment - TFDS *does* have the binarized mnist dataset. It's called `binarized_mnist` :)

Comment (jheek) [reply examples/vae/main.py]: What is your recommendation here? To eval on both?

Comment (AlexeyG) [reply examples/vae/main.py]: Just stick to the `binarized_mnist` in training and eval. This should be equivalent to the "static MNIST" setup from VAE literature.

File: examples/vae/main.py (hunk 5 above)
Comment (jheek) [reply examples/vae/main.py]: latents are a flag now 64 is because the visualization is 8x8

File: examples/vae/main.py (hunk 6 above)
Comment (jheek) [reply examples/vae/main.py]: There are a lot of reconstructions (10,000) so I think pre-processing the data for the comparison image is not a bad idea

File: examples/vae/main.py (hunk 3 above)
Comment (jheek) [reply examples/vae/main.py]: done

File: examples/vae/main.py (hunk 1 above)
Comment (jheek) [reply examples/vae/main.py]: done

File: examples/vae/main.py (hunk 2 above)
Comment (jheek) [reply examples/vae/main.py]: I like to use the simple and explicit Jax APIs as much as possible. `nn.stochastic` was introduced to avoid having to pass around rngs through a complex model. But within a single function I think manually splitting rngs is still better.

File: examples/vae/main.py (hunk 6 above)
Comment (AlexeyG) [reply examples/vae/main.py]: I didn't realise this. I thought it was going to be just the regular batch size.

File: examples/vae/main.py (hunk 1 above)
Comment (jheek) [reply examples/vae/main.py]: flattening is now part of the input pipeline

File: examples/vae/main.py [hunk 7]
```diff
@@ -127,10 +124,10 @@ def loss_fn(model):
 
 
 @jax.jit
-def eval(model, eval_ds, z):
+def eval(model, eval_ds, z, z_rng):
   xs = eval_ds['image'] / 255.0
```
[the remark this answers is not in the document]
Comment (jheek) [reply examples/vae/main.py]: also part of the input pipeline now

File: examples/vae/main.py [hunk 8]
```diff
@@ -32,18 +32,18 @@
 FLAGS = flags.FLAGS
 
 flags.DEFINE_float(
-  'learning_rate', default=1e-3,
-  help=('The leanring rate for the Adam optimizer')
+    'learning_rate', default=1e-3,
+    help=('The leanring rate for the Adam optimizer')
```
[the remark this answers is not in the document]
Comment (jheek) [reply examples/vae/main.py]: done

File: examples/vae/main.py [hunk 9]
```diff
@@ -107,14 +100,18 @@ def binary_cross_entropy_with_logits(logits, labels):
 def compute_metrics(recon_x, x, mean, logvar):
   bce_loss = binary_cross_entropy_with_logits(recon_x, x)
   kld_loss = kl_divergence(mean, logvar)
-  return {'bce': jnp.mean(bce_loss), 'kld': jnp.mean(kld_loss), 'loss': jnp.mean(bce_loss + kld_loss)}
+  return {
+      'bce': jnp.mean(bce_loss),
+      'kld': jnp.mean(kld_loss),
+      'loss': jnp.mean(bce_loss + kld_loss)
```
[the remark this answers is not in the document]
Comment (jheek) [reply examples/vae/main.py]: done
